### PR TITLE
Added unittest for Hub Module Exporter

### PR DESCRIPTION
**Pull request by captain-pool**
Initial Commit
-------------------------------
#### Tests Not Working
The Internal Test runs. Custom tests are not working.
```
[       OK ] TFHubMNISTTest.test_session
----------------------------------------------------------------------
Ran 1 test in 9.114s

OK
```
cc: @srjoglekar246 @vbardiovskyg

### CODE REVIEW

File: E1_TFHub_Sample_Deploy/export.py [hunk 1]
```diff
@@ -42,23 +49,19 @@ def train_step(model, loss_fn, optimizer_fn, metric, image, label):
...
-  label_onehot = tf.one_hot(label, 10)
-
-
-def main():
+def main(_):
   model = MNIST()
-  train, test = tfds.load("mnist", split=["train", "test"])
+  kwargs = {}
+  if FLAGS.data_dir:
+    kwargs = {"data_dir": FLAGS.data_dir}
```
Comment (srjoglekar246) [inline E1_TFHub_Sample_Deploy/export.py]: `kwargs["data_dir"] = FLAGS.data_dir`

File: E1_TFHub_Sample_Deploy/export_test.py [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+    prediction = self.model.call(sample['image'])
+    self.assertEqual(tf.argmax(prediction).numpy(), sample['label'])
+
+  def checkAccuracy(self):
+    accuracy = 0
+    for data in self.dataset:
+      preds = self.model.call(data['image'])
+      preds = tf.argmax(preds, axis=1).numpy()
+      accuracy += sum(preds == data['label'].numpy()) / self.batch
+    print("Accuracy: %.2f" % (accuracy * 100))
```
Comment (srjoglekar246) [inline E1_TFHub_Sample_Deploy/export_test.py]: Remove print statements from tests.

File: E1_TFHub_Sample_Deploy/export.py [hunk 3]
```diff
@@ -1,10 +1,17 @@
+import sys
 import tensorflow as tf
 import tensorflow_datasets as tfds
 import tensorflow_hub as hub
+from absl import app, flags
 """ Create a Sample TF-Hub Module using SavedModel v2.0
 The module has as a single signature which loads MNIST Dataset from TFDS and train a simple Neural Network for classifying the digits. The model is built and trained using Tewnsorlfow
 """
-
+FLAGS = flags.FLAGS
```
Comment (srjoglekar246) [inline E1_TFHub_Sample_Deploy/export.py]: Do not define these directly as command-line arguments. Create functions that have these arguments, which can then be called programmatically in unit tests. Then, in the `if__name__ ==...` section, parse command line flags and call the functions via main, as needed.

This enables this code to be used as a library as well as a script. Otherwise, there is no way to run it except as a command line script. Unit tests require functions to be called, they don't invoke a script.

For example, see [here](https://github.com/tensorflow/hub/blob/master/examples/text_embeddings/export.py#L180). `export_path` is an argument to a function, and then [optionally taken as a parameter](https://github.com/tensorflow/hub/blob/master/examples/text_embeddings/export.py#L262) _if_ the code is used as a script.

File: E1_TFHub_Sample_Deploy/export_test.py [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+    "/tmp/tfhub_modules/mnist/digits/1",
+    "Path to exported Module")
+flags.DEFINE_string("data_dir", None, "Path to TFDS Dataset Directory")
+flags.DEFINE_boolean("check_accuracy", False, "Checks the accuracy of a model")
+flags.DEFINE_integer("batch_size", 32, "Batch size for checking accuracy")
+flags.DEFINE_integer(
+    "buffer_size",
+    1000,
+    "Buffer size used for Data Shuffling")
+FLAGS = flags.FLAGS
```
Comment (srjoglekar246) [inline E1_TFHub_Sample_Deploy/export_test.py]: Your unit tests shouldn't contain command line flags, since this isn't a script really. Once you refactor `export.py` to have functions that take these as arguments, you will need to implement tests that call those functions with appropriate arguments.

See how `export_path` is passed as an argument [here](https://github.com/tensorflow/hub/blob/master/examples/text_embeddings/export_test.py#L67)

File: E1_TFHub_Sample_Deploy/export_test.py (hunk 2 above)
Comment (captain-pool) [reply E1_TFHub_Sample_Deploy/export_test.py]: Well, this function actually tests the accruacy of the model. So, where should I log the accuracy? I know, tests are not meant for this purpose, but I'm out of ideas for tests. 😅
Can you suggest some?

File: E1_TFHub_Sample_Deploy/export.py (hunk 3 above)
Comment (captain-pool) [reply E1_TFHub_Sample_Deploy/export.py]: Okay  :)

File: E1_TFHub_Sample_Deploy/export.py (hunk 1 above)
Comment (captain-pool) [reply E1_TFHub_Sample_Deploy/export.py]: Fixed :)

File: E1_TFHub_Sample_Deploy/export.py (hunk 3 above)
Comment (captain-pool) [reply E1_TFHub_Sample_Deploy/export.py]: Done :)

File: E1_TFHub_Sample_Deploy/export_test.py (hunk 4 above)
Comment (captain-pool) [reply E1_TFHub_Sample_Deploy/export_test.py]: Done :)

File: E1_TFHub_Sample_Deploy/export_test.py [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+
+class TFHubMNISTTest(tf.test.TestCase):
+  def setUp(self):
+    kwargs = {}
+    if FLAGS.data_dir:
+      kwargs["data_dir"] = FLAGS.data_dir
+    self.dataset = tfds.load("mnist",
+                             split="test",
+                             **kwargs).shuffle(FLAGS.buffer_size,
+                                               reshuffle_each_iteration=True).batch(FLAGS.batch_size)
```
Comment (srjoglekar246) [inline E1_TFHub_Sample_Deploy/export_test.py]: The formatting seems weird here. Can you run these files through the python linter?

File: E1_TFHub_Sample_Deploy/export_test.py (hunk 2 above)
Comment (srjoglekar246) [reply E1_TFHub_Sample_Deploy/export_test.py]: We shouldn't be loading an entire dataset in unit tests. Unit tests are usually run as a sanity check anyway.
Ideally, we should verify the outputs (via asserts) for one or two sample inputs (just put raw images in a ['testdata' folder](https://github.com/tensorflow/tensorflow/tree/master/tensorflow/core/lib/jpeg) if need be), and boundary cases such as all-zeros, invalid data, etc. There should be one test per case, like [this](https://github.com/tensorflow/hub/blob/master/examples/text_embeddings/export_test.py#L86).

File: E1_TFHub_Sample_Deploy/export.py [hunk 6]
```diff
@@ -43,22 +46,27 @@ def train_step(model, loss_fn, optimizer_fn, metric, image, label):
...
 
 
-@tf.function
-def test(image, label):
-  preds = model(image)
-  label_onehot = tf.one_hot(label, 10)
-
-
-def main():
+def train_and_export(
```
Comment (srjoglekar246) [inline E1_TFHub_Sample_Deploy/export.py]: This is a good time to add [docstrings](https://github.com/tensorflow/hub/blob/master/examples/text_embeddings/export.py#L100) to exported functions.

**Comment by srjoglekar246** [comment]
Also, we need to add a BUILD file to actually run the tests :-)

File: E1_TFHub_Sample_Deploy/export_test.py (hunk 5 above)
Comment (captain-pool) [reply E1_TFHub_Sample_Deploy/export_test.py]: I'm linting with autopep8. Let me run through pylint and check

File: E1_TFHub_Sample_Deploy/export.py (hunk 6 above)
Comment (captain-pool) [reply E1_TFHub_Sample_Deploy/export.py]: Done :)

File: E1_TFHub_Sample_Deploy/export_test.py (hunk 2 above)
Comment (captain-pool) [reply E1_TFHub_Sample_Deploy/export_test.py]: Done :)

File: E1_TFHub_Sample_Deploy/export_test.py [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+
+  def test_model_loadable(self):
+    if not os.path.exists("%s/model/1" % TMPDIR):
+      self.test_model_exporting()
+    model = load_model("%s/model/1" % TMPDIR)
+    output_ = model.call(
+        tf.zeros([1, 28, 28, 1], dtype=tf.uint8).numpy())
+    self.assertEqual(output_.shape, [1, 10])
+
+  def test_model_working(self):
```
Comment (srjoglekar246) [inline E1_TFHub_Sample_Deploy/export_test.py]: Maybe add test for empty input? Then call the tests `test_empty_input`, `test_sample_input`, etc. These test names appear in error logs (if there is a regression), so its good to have their names reflect what exactly is not working :-)

Comment (captain-pool) [reply E1_TFHub_Sample_Deploy/export_test.py]: Hey sachin, Empty inputs are checked. Using `tf.zeros`. is that what you are talking about?

Comment (srjoglekar246) [reply E1_TFHub_Sample_Deploy/export_test.py]: Aah. Didn't see that. Then rename `test_model_loadable` to `test_zero_input`, since its doing something beyond just simply loading the model.

File: E1_TFHub_Sample_Deploy/export_test.py [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+    if not os.path.exists(os.path.join(TMPDIR, "model/1")):
+      self.test_model_exporting()
+    model = load_model("%s/model/1" % TMPDIR)
+    prediction = model.call(self.test_image)
+    self.assertEqual(
+        tf.argmax(
+            tf.squeeze(prediction)).numpy(),
+        self.test_label)
```
Comment (srjoglekar246) [inline E1_TFHub_Sample_Deploy/export_test.py]: newline here.

File: E1_TFHub_Sample_Deploy/export_test.py (hunk 7 above)
Comment (captain-pool) [reply E1_TFHub_Sample_Deploy/export_test.py]: Done :)

**Comment by captain-pool** [comment]
@srjoglekar246 How about Now?? :)
I did some research on Bazel BUILD files, studied the examples you gave me, and wrote the build.
for the workspace, I copied it from tensorflow hub's workspace file.

File: E1_TFHub_Sample_Deploy/BUILD [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+    deps = ["//external:expect_numpy_installed",
+    "//external:tensorflow_hub"]
+)
+py_test(
+    name = "export_test",
+    srcs = ["export_test.py"],
+    data = glob(["test_data/*.jpg"]),
+    python_version = "PY3",
+    srcs_version = "PY2AND3",
+    deps = [":export_lib",
```
Comment (srjoglekar246) [inline E1_TFHub_Sample_Deploy/BUILD]: Formatting is weird here. Can you see the TF BUILD files for general guidelines?

File: E1_TFHub_Sample_Deploy/export_test.py (hunk 7 above)
Comment (srjoglekar246) [reply E1_TFHub_Sample_Deploy/export_test.py]: Rename this test to `test_sample_input`.

File: E1_TFHub_Sample_Deploy/export_test.py (hunk 8 above)
Comment (srjoglekar246) [reply E1_TFHub_Sample_Deploy/export_test.py]: This slipped through the cracks? :-)

Comment (captain-pool) [reply E1_TFHub_Sample_Deploy/export_test.py]: @srjoglekar246 didn't really get, what you said. where should I put the new line?

File: E1_TFHub_Sample_Deploy/BUILD (hunk 9 above)
Comment (captain-pool) [reply E1_TFHub_Sample_Deploy/BUILD]: Done :)

File: E1_TFHub_Sample_Deploy/export_test.py (hunk 7 above)
Comment (captain-pool) [reply E1_TFHub_Sample_Deploy/export_test.py]: Done :)
